File: nodes/truncate-by-tokens-preserve-system/truncate-by-tokens-preserve-system.process.js
import { encoding_for_model } from 'tiktoken';

export default async ({inputs, settings, config}) => {
    try {
        const messages = inputs.messages;
        const maxTokens = inputs.max_tokens;

        if (!Array.isArray(messages)) {
            throw new Error("Messages input must be an array");
        }

        if (typeof maxTokens !== 'number' || maxTokens < 0 || !Number.isInteger(maxTokens)) {
            throw new Error("Max tokens must be a non-negative integer");
        }

        const tokenizer = settings.tokenizer || 'cl200k_base';
        if (!['cl200k_base', 'cl100k_base'].includes(tokenizer)) {
            throw new Error(`Unsupported tokenizer: ${tokenizer}. Only cl200k_base and cl100k_base are supported.`);
        }

        const countSystem = settings.count_system === true; // Default to false

        const encoding = encoding_for_model(tokenizer === 'cl200k_base' ? 'gpt-4o' : 'gpt-4');

        // Helper function to extract text content from message
        const getMessageText = (message) => {
            if (!message || typeof message !== 'object') return '';
            if (typeof message.content === 'string') return message.content;
            if (Array.isArray(message.content)) {
                return message.content
                    .filter(item => item && item.type === 'text' && item.text)
                    .map(item => item.text)
                    .join(' ');
            }
            return String(message.content || '');
        };

        // Helper function to count tokens in a message
        const countMessageTokens = (message) => {
            const text = getMessageText(message);
            return encoding.encode(text).length;
        };

        // Separate system and non-system messages
        const systemMessages = [];
        const nonSystemMessages = [];

        for (const message of messages) {
            if (message && typeof message === 'object' && message.role === 'system') {
                systemMessages.push(message);
            } else {
                nonSystemMessages.push(message);
            }
        }

        // Count tokens in system messages
        let systemTokens = 0;
        for (const message of systemMessages) {
            systemTokens += countMessageTokens(message);
        }

        // Handle system messages based on count_system setting
        const result = [];
        let totalTokens = 0;
        let truncated = false;

        if (maxTokens === 0) {
            // Zero token limit - don't include any messages
            truncated = systemMessages.length > 0 || nonSystemMessages.length > 0;
        } else if (countSystem) {
            // System messages count toward limit - only include if they fit
            if (systemTokens <= maxTokens) {
                result.push(...systemMessages);
                totalTokens = systemTokens;
            } else {
                // System messages exceed limit - don't include them
                truncated = systemMessages.length > 0;
            }
        } else {
            // System messages don't count toward limit - always include them
            result.push(...systemMessages);
            totalTokens = systemTokens;
            if (systemTokens > maxTokens) {
                truncated = true;
            }
        }

        // Only process non-system messages if we have room and token limit > 0
        if (maxTokens > 0 && (systemTokens <= maxTokens || !countSystem)) {
            for (let i = nonSystemMessages.length - 1; i >= 0; i--) {
                const message = nonSystemMessages[i];
                const messageTokens = countMessageTokens(message);
                
                if (totalTokens + messageTokens <= maxTokens) {
                    result.push(message);
                    totalTokens += messageTokens;
                } else {
                    truncated = true;
                    break;
                }
            }
        } else if (maxTokens > 0 && systemTokens > maxTokens && countSystem) {
            // System messages exceed limit and count_system is true, so non-system messages are truncated
            truncated = nonSystemMessages.length > 0;
        }

        // Sort result to maintain original order (system messages first, then others)
        result.sort((a, b) => {
            const aIndex = messages.indexOf(a);
            const bIndex = messages.indexOf(b);
            return aIndex - bIndex;
        });

        encoding.free();

        // Count system messages in the result
        const resultSystemCount = result.filter(msg => msg && msg.role === 'system').length;

        return {
            messages: result,
            token_count: totalTokens,
            truncated: truncated,
            system_count: resultSystemCount
        };

    } catch (error) {
        console.log('Truncate by Tokens Preserve System error:', error);
        throw new Error(`Truncate by Tokens Preserve System error: ${error.message}`);
    }
};

File: nodes/truncate-by-tokens-preserve-system/truncate-by-tokens-preserve-system.process.py
import tiktoken
from typing import Any, Dict, List, Union

async def process(inputs: Dict[str, Any], settings: Dict[str, Any], config: Dict[str, Any], nodeConfig: Dict[str, Any]) -> Dict[str, Any]:
    try:
        messages = inputs.get('messages')
        max_tokens = inputs.get('max_tokens')

        if not isinstance(messages, list):
            raise Exception("Messages input must be an array")

        if not isinstance(max_tokens, int) or max_tokens < 0:
            raise Exception("Max tokens must be a non-negative integer")

        tokenizer = settings.get('tokenizer', 'cl200k_base')
        if tokenizer not in ['cl200k_base', 'cl100k_base']:
            raise Exception(f"Unsupported tokenizer: {tokenizer}. Only cl200k_base and cl100k_base are supported.")

        count_system = settings.get('count_system', False)  # Default to False

        encoding_name = 'cl200k_base' if tokenizer == 'cl200k_base' else 'cl100k_base'
        encoding = tiktoken.get_encoding(encoding_name)

        # Helper function to extract text content from message
        def get_message_text(message):
            if not isinstance(message, dict):
                return ''
            content = message.get('content', '')
            if isinstance(content, str):
                return content
            elif isinstance(content, list):
                text_parts = []
                for item in content:
                    if isinstance(item, dict) and item.get('type') == 'text' and 'text' in item:
                        text_parts.append(item['text'])
                return ' '.join(text_parts)
            else:
                return str(content)

        # Helper function to count tokens in a message
        def count_message_tokens(message):
            text = get_message_text(message)
            return len(encoding.encode(text))

        # Separate system and non-system messages
        system_messages = []
        non_system_messages = []

        for message in messages:
            if isinstance(message, dict) and message.get('role') == 'system':
                system_messages.append(message)
            else:
                non_system_messages.append(message)

        # Count tokens in system messages
        system_tokens = 0
        for message in system_messages:
            system_tokens += count_message_tokens(message)

        # Handle system messages based on count_system setting
        result = []
        total_tokens = 0
        truncated = False

        if max_tokens == 0:
            # Zero token limit - don't include any messages
            truncated = len(system_messages) > 0 or len(non_system_messages) > 0
        elif count_system:
            # System messages count toward limit - only include if they fit
            if system_tokens <= max_tokens:
                result.extend(system_messages)
                total_tokens = system_tokens
            else:
                # System messages exceed limit - don't include them
                truncated = len(system_messages) > 0
        else:
            # System messages don't count toward limit - always include them
            result.extend(system_messages)
            total_tokens = system_tokens
            if system_tokens > max_tokens:
                truncated = True

        # Only process non-system messages if we have room and token limit > 0
        if max_tokens > 0 and (system_tokens <= max_tokens or not count_system):
            for i in range(len(non_system_messages) - 1, -1, -1):
                message = non_system_messages[i]
                message_tokens = count_message_tokens(message)
                
                if total_tokens + message_tokens <= max_tokens:
                    result.append(message)
                    total_tokens += message_tokens
                else:
                    truncated = True
                    break
        elif max_tokens > 0 and system_tokens > max_tokens and count_system:
            # System messages exceed limit and count_system is true, so non-system messages are truncated
            truncated = len(non_system_messages) > 0

        # Sort result to maintain original order (system messages first, then others)
        result.sort(key=lambda x: messages.index(x) if x in messages else 0)

        # Count system messages in the result
        result_system_count = len([msg for msg in result if isinstance(msg, dict) and msg.get('role') == 'system'])

        return {
            "messages": result,
            "token_count": total_tokens,
            "truncated": truncated,
            "system_count": result_system_count
        }

    except Exception as error:
        print(f'Truncate by Tokens Preserve System error: {error}')
        raise Exception(f"Truncate by Tokens Preserve System error: {str(error)}")
